### DISCUSSION TITLE
Cannot save output data (tried several ways)

### DISCUSSION THREAD

**Discussion by Altr37** [jspsych/jsPsych#1677] [Q&A] [answered]
Hi,

I have no experience in coding, and went thoroughly through the guides, and tried several different ways of saving, either locally or to the server, and nothing worked, and I'm not sure why.
Things I tried:
(1) 
```js
jsPsych.init({
    timeline: timeline,
    on_finish: function() {
      jsPsych.data.get().filter({trial_part: 'probe'||'question', trial_type: 'html-keyboard-response'}).csv();
    }
})
```
* doesn't show an error, but not file created

(2)  
```js
var all_data = jsPsych.data.get().filter({trial_part: 'probe'||'question', trial_type: 'html-keyboard-response'});
      console.log(all_data.csv());
      console.log(jsPsych.data.get().csv());
     jsPsych.data.get().localSave('csv','all_data.csv');
```
* last line produces an error: 'cannot read property 'InsertAdjacentHTML'

(3) created notepad file, name ending with PHP:
```php
<?php
$post_data = json_decode(file_get_contents('php://input'), true);
// the directory "data" must be writable by the server
$name = "data/".$post_data['filename'].".csv";
$data = $post_data['filedata'];
// write the file to disk
file_put_contents($name, $data);
?>
```
then added this line in my experiment HTML under <>body<>
      `<?php include 'write_data.php';?>`

and then
```js
function saveData(name, data){
  var xhr = new XMLHttpRequest();
  xhr.open('POST', 'write_data.php'); // 'write_data.php' is the path to the php file described above.
  xhr.setRequestHeader('Content-Type', 'application/json');
  xhr.send(JSON.stringify({filename: name, filedata: data}));
}

// call the saveData function after the experiment is over
jsPsych.init({
  timeline: timeline,
   // code to define the experiment structure would go here...
   on_finish: function(){ saveData("experiment_data", jsPsych.data.get().filter({trial_part: 'probe'||'question', trial_type: 'html-keyboard-response'}).csv());
   }
});
```
* doesn't produce any file, but no error.

Important to note: if I just ask to display the data on the screen at the end, it does so, though without applying the filter I asked for.

Is there anyone who can help me? 
I want to save filtered data ({trial_part: 'probe'||'question', trial_type: 'html-keyboard-response'})) in csv format, in /data folder on the server, under subject_id name (var subject_id = jsPsych.randomization.randomID(15);)

Thank you all, sorry for being too long.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @Altr37, I'm not sure if this is the only issue, but one problem with this is that you can't use the OR (||) operator inside of jsPsych's `filter` function. If you want to filter on one of two possible values, then you can use the [`filterCustom` function](https://www.jspsych.org/core_library/jspsych-data/#filtercustom), which allows you to add in logical operators, like this:
```js
var data = jsPsych.data.get().filterCustom(function(trial){
    return (trial.trial_type == "html-keyboard-response" && (trial.trial_part = "probe" || trial.trial_part = "question"));
});
```
Then you can try printing the result to the console, to make sure that it's working:
```js
var data = jsPsych.data.get().filterCustom(function(trial){
    return (trial.trial_type == "html-keyboard-response" && (trial.trial_part = "probe" || trial.trial_part = "question"));
});
// print the raw values
console.log(data.values());
// print the CSV-formatted data
console.log(data.csv());
```
You can also try using the `localSave` function. You need to call `localSave` on the filtered data collection that you just created:
```js
var data = jsPsych.data.get().filterCustom(function(trial){
    return (trial.trial_type == "html-keyboard-response" && (trial.trial_part = "probe" || trial.trial_part = "question"));
});
data.localSave('csv','mydata.csv');
```
As I said I'm not sure if there are other problems with your code. But I would start by making sure to get this data filter part working, and after that, try saving the data to the server.

**Reply by Altr37** [reply]
Thank you @becky-gilbert  for your reply.
Unfortunately, it didn't work.
The filter line produced an error: invalid assignment left-hand side
I changed the phrase so it will only contain == like this:

var data = jsPsych.data.get().filterCustom(function(trial){
    return (trial.trial_type == "html-keyboard-response" && (trial.trial_part == "probe" || trial.trial_part == "question"));
});

Then it was able to run the experiment, but there was no output on the console.

In addition, if I include the data.localSave('csv','mydata.csv') it doesn't run the experiment at all, and produces an error: 
jspsych.js:1614 Uncaught TypeError: Cannot read property 'insertAdjacentHTML' of undefined

Any ideas on how to approach this?

Thank you.

**Reply by becky-gilbert** [reply]
Ah good catch about the = vs == mistake in my response. Sorry about that!

I'm not sure why this isn't working. Can you let me know what version of jsPsych you're using, so that I can try to reproduce it?

**Reply by Altr37** [reply]
Thanks for the quick reply!
I'm using 6.2.0 version.

It seems that this filter still doesn't work, as it shows no output on the console, and shows blank screen when I try to present it as a var, like this:
  var data_subset = jsPsych.data.get().filterCustom(function(trial){
  return (trial.trial_type == "html-keyboard-response" && ((trial.trial_part == "probe") || (trial.trial_part == "question")));
  });
   var datasum= {
      type: 'html-keyboard-response',
      stimulus: data_subset,
      choices: ['space']
};

timeline.push(datasum);
 
It just shows on the screen: [object]

The other thing is that local save doesn't work, regardless of the filter (even if I try to show all data), like:

jsPsych.data.get().localSave('csv','mydata.csv');

 it produces an error:
Cannot read property 'insertAdjacentHTML' of undefined

**Reply by becky-gilbert** [reply]
Hi @Altr37, I tried to replicate this using v6.2.0 and it's working fine for me. Here's my code:
```js
  var timeline = [];
  var start = {
    type: 'html-keyboard-response',
    stimulus: "Press any key to start.",
    data: {trial_part: "start"}
  };
  timeline.push(start);
  var probe = {
    type: 'html-keyboard-response',
    stimulus: jsPsych.timelineVariable('probe'),
    data: {trial_part: "probe"}
  };
  var question = {
    type: 'html-keyboard-response',
    stimulus: jsPsych.timelineVariable('question'),
    data: {trial_part: "question"}
  };
  var procedure = {
    timeline: [probe, question],
    timeline_variables: [
      {probe: "1", question: "Q1"},
      {probe: "2", question: "Q2"},
      {probe: "3", question: "Q3"},
    ]
  };
  timeline.push(procedure);
  var end = {
    type: 'html-keyboard-response',
    stimulus: "Done!",
    data: {trial_part: "end"}
  };
  timeline.push(end);

  jsPsych.init({
    timeline: timeline,
    on_finish: function () {
      var filtered_data = jsPsych.data.get().filterCustom(function(trial) {
        return (trial.trial_type == "html-keyboard-response" && ((trial.trial_part == "probe") || (trial.trial_part == "question")));
      });
      filtered_data.localSave('csv','filtered_data.csv');
    }
  });
```
And here's the CSV, which just contains the probe and question trials:
![filtered_data](https://user-images.githubusercontent.com/9041788/113191033-ea6c5680-9211-11eb-8a07-55867d92445a.png)

I'm not sure why the `localSave` function isn't working for you. Can you try upgrading to a newer browser version or switching to a different browser? 

Or maybe there's some other issue with your code. What happens when you try my version - does it work for you?

By the way, the reason that your `datasum` trial didn't work is because the `data_subset` variable is in the main script, so it will be created when the exeperiment first loads, at which point there is no jsPsych data. So to get the `data_subset` variable to be created _during_ the experiment and used as a trial stimulus, you need to make the stimulus parameter into a [dynamic parameter](https://www.jspsych.org/overview/dynamic-parameters/), i.e. a function.  You also need to turn the object into a string so that it can be printed on the screen (otherwise you will always see [object]). One way to do that is to call [jsPsych's `json()` function](https://www.jspsych.org/core_library/jspsych-data/#json) on the data collection. This version works of your `datasum` trial works for me:
```js
var datasum = {
  type: 'html-keyboard-response',
  stimulus: function() {
    var data_subset = jsPsych.data.get().filterCustom(function(trial){
      return (trial.trial_type == "html-keyboard-response" && ((trial.trial_part == "probe") || (trial.trial_part == "question")));
    });
    return data_subset.json();
  },
  choices: ['space']
};
timeline.push(datasum);
```

**Reply by Altr37** [reply]
Thank you so much!
Your code works! When I tried to apply it on my experiment I it saved some data but only for the probes.
I think I understand what might be the problem. I have two different types of stimuli, sentences and word pairs:
var sent_v1 = [
    {
        sentence: "It required 4 men to pick up the anchor.",
        question: "Were 3 men enough to pick up the anchor?",
        correct_response: "f",
        condition: "v1",
        task_part: "sentences"
    }

var pairs_v1 = [
    {
        target: "anchor",
        probe: "luggage",
        correct_response: "j",
        condition: "con",
        task_part: "pairs"
    }

I was trying to save subjects' interaction with the question in part1, and with the probe in part 2.
With your code example it saves a csv, but only collects results for the probes. Is there a way to ask it to collect data for both the question and the probe?

I was also wondering how can I save the data under a changing name: randomly generated subject id?

**Reply by becky-gilbert** [reply]
Hi @Altr37, if I understand correctly then I think you can just filter on task_part = "pairs" or task_part = "sentences", which would look like this:
```js
jsPsych.init({
    timeline: timeline,
    on_finish: function () {
      var filtered_data = jsPsych.data.get().filterCustom(function(trial) {
        return (trial.task_part== "pairs") || (trial.task_part == "sentences");
      });
      filtered_data.localSave('csv','filtered_data.csv');
    }
  });
```
Note that this only works if you're adding `task_part` to the trial data. I can't tell for sure without seeing how you set up the trials. I'm guessing you're using timeline variables, so assuming that's the case, you just need to be sure that you're grabbing `task_part` from the timeline variables and adding it to the trial data, e.g.
```js
var trial = {
  // other trial parameters 
  data: {task_part: jsPsych.timelineVariable('task_part')}
};
```

> I was also wondering how can I save the data under a changing name: randomly generated subject id?

This is different from your original question so I've answered this in a separate post here: #1699 just to make it easier for others to find 😃

**Reply by Altr37** [reply]
THank you so much! This has been very helpful!
